Bolus progress

ps2 opened the pull request.

ps2 committed:
Fixes for mock pumpmanager

Implement PumpManager protocol required properties that interleaved before the merge.
Fix saving basal schedules and delivery limits
Fix cancel button on pump setup

ps2 committed:
PumpManager protocol changes to support bolus progress and bolus cancellation

ps2 committed:
Delivery estimator for mock pump

ps2 committed:
Merge remote-tracking branch 'tidepool/dev' into bolus-progress

ps2 committed:
Support multiple observers of dose progress

ps2 committed:
Notify outside of lock

mpangburn commented on MockKit/MockDoseProgressEstimator.swift:
```diff
@@ -0,0 +1,96 @@
+import LoopKit
+
+class MockDoseProgressEstimator: DoseProgressEstimator {
+
```
Nit: extraneous line break

mpangburn commented on MockKit/MockDoseProgressEstimator.swift:
```diff
@@ -0,0 +1,96 @@
+        os_unfair_lock_lock(&lock)
+        defer {
+            os_unfair_lock_unlock(&lock)
+        }
```
[An article I came across recently](http://www.russbishop.net/the-law) describes how Swift's exclusivity model makes the expected use of os_unfair_lock unsafe. We should favor NSLock or DispatchQueue.

ps2 replied:
Added new UnfairLock, and updated existing Lock to deal with potential exclusivity enforcement.

darinkrauss commented on LoopKit/DoseProgressEstimator.swift:
```diff
@@ -0,0 +1,36 @@
+
+import Foundation
+
+public struct DoseProgress {
```
Would it be useful/logical to include the actual DoseEntry here? I note that the two implementations of DoseProgressEstimator (MockDoseProgressEstimator and MinimedDoseProgressEstimator) both have DoseEntry.

darinkrauss replied:
That way, there is no ambiguity about what does is in progress (and you don't need to guess the total dose if you wanted to display that).

ps2 replied:
I'd like to minimize the contract surface area of this protocol, and also limit redundancy in what PumpManager vends.

darinkrauss replied:
👍

darinkrauss commented on MockKit/MockDoseProgressEstimator.swift:
```diff
@@ -0,0 +1,96 @@
+import Foundation
+import LoopKit
+
+class MockDoseProgressEstimator: DoseProgressEstimator {
```
Overall, might want to make MockDoseProgressEstimator match much of MinimedDoseProgressEstimator. In fact, they are similar enough that one might imagine a base class implementation. Just a thought.

darinkrauss replied:
Also, use of lock differs between the two implementations.

ps2 replied:
Updated to use a base class.

mpangburn reviewed (approved):
LGTM! Only one real note, re locking.

darinkrauss reviewed (changes requested):
A couple thoughts. Primarily, though, differences between the Mock* and Minimed* versions of these types.

ps2 committed:
Updates from reviews:

- Make Lock classes public
- Update locks for potential exclusivity enforcement
- Base class for dose progress estimation

ps2 committed:
Renaming

ps2 committed:
Use DispatchSourceTimer instead of running on main loop

ps2 committed:
Create bolus progress reporter on demand, and report on specified queue

ps2 committed:
Stop timer when progress is complete.

mpangburn commented on LoopKit/DoseProgressTimerEstimator.swift:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public func notify() {
+        let observersCopy = lock.locked { () -> WeakSet<DoseProgressObserver> in
```
You should be able to remove the type annotation here, i.e. `lock.locked { observers }`

mpangburn commented on LoopKit/DoseProgressTimerEstimator.swift:
```diff
@@ -0,0 +1,96 @@
+            return
+        }
+
+        timer.setEventHandler {}
```
What's the effect of removing the event handler right before canceling the timer?

ps2 replied:
Perhaps it is unnecessary, but I saw that usage here: https://medium.com/over-engineering/a-background-repeating-timer-in-swift-412cecfd2ef9, which covered some important details about setting up and deallocating GCD timers.

mpangburn commented on LoopKit/Locked.swift:
```diff
@@ -0,0 +1,44 @@
+
+
+public class Locked<T> {
+    private var _lock: UnsafeMutablePointer<os_unfair_lock>
```
Might be simpler to implement Locked in terms of UnfairLock to deduplicate the initialization logic here.

mpangburn commented on LoopKit/UnfairLock.swift:
```diff
@@ -0,0 +1,38 @@
+        _lock.deallocate()
+    }
+
+    public func locked<ReturnValue>(_ f: () throws -> ReturnValue) rethrows -> ReturnValue {
```
`withLock` reads more clearly IMO.

mpangburn commented on MockKit/Extensions/Collection.swift:
```diff
@@ -7,6 +7,7 @@
 //
 
 import Dispatch
+import LoopKit
```
This import addition shouldn't be necessary; the extension method in this file is agnostic to LoopKit data types.

ps2 replied:
I made Locked public in LoopKit, since we're using it it a lot of places that use LoopKit.

mpangburn reviewed (approved):
🚀

ps2 committed:
Updates from reviews:

- Use UnfairLock in Locked implementation
- locked -> withLock
- Prevent unsafe access of timer during deinit
- brevity tweaks

darinkrauss commented on LoopKit.xcodeproj/project.pbxproj:
```diff
@@ -1316,6 +1324,7 @@
 				43D8FDED1C7290350073BE78 /* HealthKitSampleStore.swift */,
 				43B17C88208EEC0B00AC27E9 /* HealthStoreUnitCache.swift */,
 				43F5034A21051FCD009FA89A /* KeychainManager.swift */,
+				4353D179203E7840007B4ECD /* Locked.swift */,
```
Is `Locked.swift` still included in the project somewhere?

darinkrauss replied:
Scratch that. I see that it just got moved to a different folder.

darinkrauss commented on LoopKit/DoseProgressTimerEstimator.swift:
```diff
@@ -0,0 +1,94 @@
+        self.reportingQueue = reportingQueue
+    }
+
+    open var progress: DoseProgress {
```
Couldn't you, in theory, remove this here and remove the `DoseProgressReporter` protocol above, and place both on the subclasses?

darinkrauss commented on MockKit/MockDoseProgressEstimator.swift:
```diff
@@ -0,0 +1,44 @@
+        case .bolus:
+            timeBetweenPulses = TimeInterval(2)
+        case .basal, .tempBasal:
+            timeBetweenPulses = 0.05 / dose.unitsPerHour
```
This doesn't look right. If the basal dose is 2 U/hr, then `timeBetweenPulses` is .025. There is a pulse every 1/40th of a second?

darinkrauss replied:
Maybe multiply by 60*60 so it is per hour?

ps2 replied:
Thanks. Yep!

darinkrauss reviewed (changes requested):
Only one real concern with the TimeInterval for basals. The math may be wrong. If not, let me know and I'll approve.

ps2 committed:
Fix timeBetweenPulses calc for basal rates

darinkrauss reviewed (approved):
LGTM 🚀
